File: code/chapter3/linear_regression.py
import numpy as np


class LR:
    def __init__(self):
        self.coef = None

    def qr_solver(self, x, y):
        q, r = np.linalg.qr(x)
        p = np.dot(q.T, y)
        return np.dot(np.linalg.inv(r), p)

    def fit(self, x, y):
        self.coef = self.qr_solver(np.hstack((np.ones((x.shape[0], 1)), x)), y)

    def predict(self, x):
        return np.dot(np.hstack((np.ones((x.shape[0], 1)), x)), self.coef)

File: code/chapter4/linear_regression.R

library(R6)
LR = R6Class(
  "LR",
  public = list(
    coef = NULL,
    initialize = function() {
      
    },
    fit = function(x, y) {
      self$qr_solver(cbind(1, x), y)
    },
    qr_solver = function(x, y) {
      self$coef = qr.coef(qr(x), y)
    },
    predict = function(new_x) {
      cbind(1, new_x) %*% self$coef
    }
  )
)
